### PR TITLE
feat!(scalasdk): Public API for Value Entities.

### PR DESCRIPTION
**Pull request by RayRoestenburg**
First cut of Public API for Value Entities, lot's of wrapping and unwrapping.

### CODE REVIEW

File: sdk/scala-sdk/src/main/scala/com/akkaserverless/scalasdk/ServiceCallRef.scala [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+package com.akkaserverless.scalasdk
+
+import com.akkaserverless.javasdk
+import com.google.protobuf.Descriptors
+
+/**
+ * A reference to a call on a service.
+ *
+ * @param <T>
```
Comment (octonato) [inline sdk/scala-sdk/src/main/scala/com/akkaserverless/scalasdk/ServiceCallRef.scala:25]:
```suggestion
 * @param [T]
```

File: sdk/scala-sdk/src/main/scala/com/akkaserverless/scalasdk/impl/valueentity/ValueEntityAdapter.scala [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+import com.akkaserverless.scalasdk.impl.MetadataImpl
+import com.akkaserverless.scalasdk.impl.valueentity.ValueEntityEffectImpl
+import com.akkaserverless.scalasdk.valueentity.CommandContext
+import com.akkaserverless.scalasdk.valueentity.ValueEntity
+import com.akkaserverless.scalasdk.valueentity.ValueEntityContext
+import com.akkaserverless.scalasdk.valueentity.ValueEntityOptions
+import com.akkaserverless.scalasdk.valueentity.ValueEntityProvider
+import com.google.protobuf.Descriptors
+
+private[scalasdk] class Scala2JavaValueEntityAdapter[S](scalasdkValueEntity: ValueEntity[S])
```
Comment (octonato) [inline sdk/scala-sdk/src/main/scala/com/akkaserverless/scalasdk/impl/valueentity/ValueEntityAdapter.scala:35]: On #433, I have used Scala2Java and Java2Scala for the Action adapter, but also introduced `JavaServiceCallAdapter`. The language prefix indicating that it adapts to Java. 

I think we can remove `Scala2` and `Java2`. Not on that PR, but nicer if we align later.

**Review by octonato — APPROVED** [review]
LGTM

File: sdk/scala-sdk/src/main/scala/com/akkaserverless/scalasdk/Context.scala [hunk 3]
```diff
@@ -16,5 +16,5 @@
 
 package com.akkaserverless.scalasdk
 
-//FIXME
+//FIXME (not extend from javasdk Context?)
```
Comment (patriknw) [inline sdk/scala-sdk/src/main/scala/com/akkaserverless/scalasdk/Context.scala:19]: yes, I guess that shouldn't be needed since the adapters create new instances, but can be cleanup in another PR

File: sdk/scala-sdk/src/main/scala/com/akkaserverless/scalasdk/EntityOptions.scala [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.akkaserverless.scalasdk
+
+import com.akkaserverless.javasdk.impl.ComponentOptions
+
+/** Options used for configuring an entity. */
+trait EntityOptions extends ComponentOptions {
```
Comment (patriknw) [inline sdk/scala-sdk/src/main/scala/com/akkaserverless/scalasdk/EntityOptions.scala:22]: add the explicit methods for forward headers?

File: sdk/scala-sdk/src/main/scala/com/akkaserverless/scalasdk/impl/valueentity/ValueEntityAdapter.scala (hunk 2 above)
Comment (RayRoestenburg) [reply sdk/scala-sdk/src/main/scala/com/akkaserverless/scalasdk/impl/valueentity/ValueEntityAdapter.scala:35]: yes that is a little easier to read

File: sdk/scala-sdk/src/main/scala/com/akkaserverless/scalasdk/PassivationStrategy.scala [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+/** A passivation strategy. */
+trait PassivationStrategy {
+
+  /**
+   * Create a passivation strategy that passivates the entity after the default duration (30 seconds) of inactivity.
+   *
+   * @return
+   *   the passivation strategy
+   */
+  def defaultTimeout = new Timeout()
```
Comment (patriknw) [inline sdk/scala-sdk/src/main/scala/com/akkaserverless/scalasdk/PassivationStrategy.scala:35]: `javasdk.impl.Timeout` shouldn't be exposed in public api

File: sdk/scala-sdk/src/main/scala/com/akkaserverless/scalasdk/PassivationStrategy.scala [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+  /**
+   * Create a passivation strategy that passivates the entity after a given duration of inactivity.
+   *
+   * @param duration
+   *   of inactivity after which the passivation should occur.
+   * @return
+   *   the passivation strategy
+   */
+  def timeout(duration: FiniteDuration) = new Timeout(duration.toJava)
```
Comment (patriknw) [inline sdk/scala-sdk/src/main/scala/com/akkaserverless/scalasdk/PassivationStrategy.scala:45]: explicit return type on all public api

File: sdk/scala-sdk/src/main/scala/com/akkaserverless/scalasdk/ServiceCallRef.scala (hunk 1 above)
Comment (patriknw) [reply sdk/scala-sdk/src/main/scala/com/akkaserverless/scalasdk/ServiceCallRef.scala:25]: `tparam` even

File: sdk/scala-sdk/src/main/scala/com/akkaserverless/scalasdk/impl/valueentity/ValueEntityAdapter.scala (hunk 2 above)
Comment (patriknw) [reply sdk/scala-sdk/src/main/scala/com/akkaserverless/scalasdk/impl/valueentity/ValueEntityAdapter.scala:35]: > The language prefix indicating that it adapts to ...

yes, that would be enough

File: sdk/scala-sdk/src/main/scala/com/akkaserverless/scalasdk/impl/valueentity/ValueEntityAdapter.scala [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+    scalasdkHandler: ValueEntityHandler[S, ValueEntity[S]])
+    extends javasdk.impl.valueentity.ValueEntityHandler[S, javasdk.valueentity.ValueEntity[S]](javasdkValueEntity) {
+
+  override def handleCommand(
+      commandName: String,
+      state: S,
+      command: Any,
+      context: javasdk.valueentity.CommandContext): javasdk.valueentity.ValueEntity.Effect[_] = {
+    scalasdkHandler.handleCommand(commandName, state, command, new Scala2JavaCommandContextAdapter(context)) match {
+      case ValueEntityEffectImpl(javasdkEffectImpl) => javasdkEffectImpl
```
Comment (patriknw) [inline sdk/scala-sdk/src/main/scala/com/akkaserverless/scalasdk/impl/valueentity/ValueEntityAdapter.scala:73]: nice

File: sdk/scala-sdk/src/main/scala/com/akkaserverless/scalasdk/impl/valueentity/ValueEntityEffectImpl.scala [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+object ValueEntityEffectImpl {
+  def apply[S](): ValueEntityEffectImpl[S] = ValueEntityEffectImpl(
+    new javasdk.impl.valueentity.ValueEntityEffectImpl[S]())
+}
+
+case class ValueEntityEffectImpl[S](javasdkEffect: javasdk.impl.valueentity.ValueEntityEffectImpl[S])
+    extends ValueEntity.Effect.Builder[S]
+    with ValueEntity.Effect.OnSuccessBuilder[S]
+    with ValueEntity.Effect[S] {
+  def deleteState: ValueEntity.Effect.OnSuccessBuilder[S] = new ValueEntityEffectImpl(javasdkEffect.deleteState())
```
Comment (patriknw) [inline sdk/scala-sdk/src/main/scala/com/akkaserverless/scalasdk/impl/valueentity/ValueEntityEffectImpl.scala:36]: that was a good solution

File: sdk/scala-sdk/src/main/scala/com/akkaserverless/scalasdk/valueentity/ValueEntity.scala [hunk 9]
```diff
@@ -16,86 +16,171 @@
...
-   * The name of the command being executed.
-   *
-   * @return
-   *   The name of the command.
-   */
-  def commandName: String
+    trait OnSuccessBuilder[S] {
+
+      /**
+       * Reply after for example <code>updateState</code>.
```
Comment (patriknw) [inline sdk/scala-sdk/src/main/scala/com/akkaserverless/scalasdk/valueentity/ValueEntity.scala:110]: for ScalaDoc we can use backtick

File: sdk/scala-sdk/src/main/scala/com/akkaserverless/scalasdk/valueentity/ValueEntity.scala [hunk 10]
```diff
@@ -16,86 +16,171 @@
...
+    }
+  }
 
   /**
-   * The id of the command being executed.
+   * A return type to allow returning forwards or failures, and attaching effects to messages.
    *
-   * @return
-   *   The id of the command.
+   * @param [T]
```
Comment (patriknw) [inline sdk/scala-sdk/src/main/scala/com/akkaserverless/scalasdk/valueentity/ValueEntity.scala:164]: `@tparam`

**Review by patriknw — APPROVED** [review]
LGTM, a few minor comments for follow up

File: sdk/scala-sdk/src/main/scala/com/akkaserverless/scalasdk/impl/valueentity/ValueEntityEffectImpl.scala (hunk 8 above)
Comment (RayRoestenburg) [reply sdk/scala-sdk/src/main/scala/com/akkaserverless/scalasdk/impl/valueentity/ValueEntityEffectImpl.scala:36]: thanks!
